### PR TITLE
`service create` beginning implementation

### PR DESCRIPTION
**Pull request by sixolet**
This implements `kn service create` supporting only the `--image` and `--env` flags.

It also provides a structure for collecting flags that will be in common between `create`, `update`, and `replace`, and using them to modify a Configuration (so they can also be reused for `configuration create` etc later)

### CODE REVIEW

File: pkg/kn/commands/configuration_edit_flags.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+)
+
+type ConfigurationEditFlags struct {
+	Image string
+	Env   map[string]string
+}
+
+func (p *ConfigurationEditFlags) AddFlags(command *cobra.Command) {
+	command.Flags().StringVar(&p.Image, "image", "", "Image to run.")
+	command.Flags().StringToStringVar(&p.Env, "env", map[string]string{},
```
Comment (cppforlife) [inline pkg/kn/commands/configuration_edit_flags.go]: comma can be present in the values: https://github.com/cppforlife/knctl/issues/58

File: pkg/kn/commands/configuration_edit_flags.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package commands
+
+import (
+	serving_lib "github.com/knative/client/pkg/serving"
```
Comment (cppforlife) [inline pkg/kn/commands/configuration_edit_flags.go]: it's not typical to have underscores in pkg aliases in go

File: pkg/kn/commands/configuration_edit_flags.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+func (p *ConfigurationEditFlags) AddFlags(command *cobra.Command) {
+	command.Flags().StringVar(&p.Image, "image", "", "Image to run.")
+	command.Flags().StringToStringVar(&p.Env, "env", map[string]string{},
+		"Environment, comma-separated NAME=value.")
+}
+
+func (p *ConfigurationEditFlags) Apply(config *servingv1alpha1.ConfigurationSpec) (err error) {
+	err = serving_lib.UpdateEnvVars(config, p.Env)
```
Comment (cppforlife) [inline pkg/kn/commands/configuration_edit_flags.go]: can throw in `err :=` and kill off named return

File: pkg/kn/commands/service_create.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	servingv1alpha1 "github.com/knative/serving/pkg/apis/serving/v1alpha1"
+
+	serving_lib "github.com/knative/client/pkg/serving"
+	"github.com/spf13/cobra"
+)
+
+func NewServiceCreateCommand(p *KnParams) *cobra.Command {
+
```
Comment (cppforlife) [inline pkg/kn/commands/service_create.go]: extra newline.

File: pkg/kn/commands/service_create.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+			namespace := cmd.Flag("namespace").Value.String()
+
+			service := servingv1alpha1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      args[0],
+					Namespace: namespace,
+				},
+			}
+			service.Spec.RunLatest = &servingv1alpha1.RunLatestType{}
+			service.GetObjectKind().SetGroupVersionKind(schema.GroupVersionKind{
```
Comment (cppforlife) [inline pkg/kn/commands/service_create.go]: dont think this is (`SetGroupVersionKind(...)`) necessary as object is typed as servingv1alpha1.Service

File: pkg/serving/config_changes.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	servingv1alpha1 "github.com/knative/serving/pkg/apis/serving/v1alpha1"
+)
+
+// Give the configuration all the env var values listed in the given map of
+// vars.  Does not touch any environment variables not mentioned, but it can add
+// new env vars and change the values of existing ones.
+func UpdateEnvVars(config *servingv1alpha1.ConfigurationSpec, vars map[string]string) error {
+	set := make(map[string]bool)
+	for i, _ := range config.RevisionTemplate.Spec.Container.Env {
+		env_var := &config.RevisionTemplate.Spec.Container.Env[i]
```
Comment (cppforlife) [inline pkg/serving/config_changes.go]: camel case is preferred

Comment (sixolet) [reply pkg/serving/config_changes.go]: oops. Code switching too fast from py.

File: pkg/serving/config_changes.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+		value, present := vars[env_var.Name]
+		if present {
+			env_var.Value = value
+			set[env_var.Name] = true
```
Comment (cppforlife) [inline pkg/serving/config_changes.go]: TODO assumes that env var name is not empty

File: pkg/serving/config_changes.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+				})
+		}
+	}
+	return nil
+
+}
+
+// Utility function to translate between the API list form of env vars, and the
+// more convenient map form.
+func EnvToMap(vars []corev1.EnvVar) (map[string]string, error) {
```
Comment (cppforlife) [inline pkg/serving/config_changes.go]: seems like this could be moved into test file.

File: pkg/kn/commands/configuration_edit_flags.go (hunk 1 above)
Comment (sixolet) [reply pkg/kn/commands/configuration_edit_flags.go]: Oh, damn. Somehow I expected this very-well-used flags lib to support some kind of quoting, but I guess not. Ugh, I'll use the StringArray version.

File: pkg/serving/config_changes.go (hunk 8 above)
Comment (sixolet) [reply pkg/serving/config_changes.go]: It's useful enough I kinda want to leave it lib for now.

File: pkg/kn/commands/service_create.go [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+func NewServiceCreateCommand(p *KnParams) *cobra.Command {
+
+	var editFlags ConfigurationEditFlags
+
+	serviceCreateCommand := &cobra.Command{
+		Use:   "create NAME",
+		Short: "Create a service.",
+		RunE: func(cmd *cobra.Command, args []string) (err error) {
+			if len(args) < 1 {
```
Comment (cppforlife) [inline pkg/kn/commands/service_create.go]: `!=` instead of `<` ?

File: pkg/kn/commands/configuration_edit_flags.go (hunk 3 above)
Comment (sixolet) [reply pkg/kn/commands/configuration_edit_flags.go]: Ok.

File: pkg/kn/commands/service_create.go (hunk 5 above)
Comment (sixolet) [reply pkg/kn/commands/service_create.go]: Oh hey you're right.

File: pkg/serving/config_changes.go (hunk 7 above)
Comment (sixolet) [reply pkg/serving/config_changes.go]: Can you explain? I think it works even with env var name empty, no?

**Pull request merged**